backend/tag/views.py

```diff
@@ -1,27 +1,139 @@
 from rest_framework.response import Response
 from rest_framework.viewsets import ModelViewSet
+from rest_framework.views import APIView
+from django.http import HttpRequest
+from rest_framework.exceptions import NotFound, PermissionDenied, ParseError
 
-from .models import Tag
+from .models import Tag, TaggedDocument
 from .permissions import IsTagOwner
 from .serializers import TagSerializer
+from addcorpus.models import Corpus
+from addcorpus.permissions import CorpusAccessPermission
 
+def check_corpus_name(request: HttpRequest):
+    '''
+    Returns the name of the corpus specified in the request query parameters,
+    if there is one.
+
+    Raises 404 if this is not a real corpus.
+    '''
+
+    corpus_name = request.query_params.get('corpus', None)
+    if corpus_name and not Corpus.objects.filter(name=corpus_name):
+        raise NotFound(f'corpus {corpus_name} does not exist')
+    return corpus_name
 
 class TagViewSet(ModelViewSet):
     serializer_class = TagSerializer
     permission_classes = [IsAuthenticated, IsTagOwner]
     queryset = Tag.objects.all()
 
     def perform_create(self, serializer):
-        '''Overwrites ModelViewSet.perform_create
+        '''Overwrites ModelViewSet.perfor0m_create
         Auto-assigns the authenticated user on creation'''
         return serializer.save(user=self.request.user)
 
     def list(self, *args, **kwargs):
         '''Overwrites ModelViewSet.list
         Filters the default queryset by ownership.
         Only applies to list view, the class queryset is unaffected.
+
+        Supports filtering on a corpus by specifying the name as a query parameter.
         '''
-        queryset = self.queryset.filter(user=self.request.user)
+
+
+        filters = {
+            'user': self.request.user,
+        }
+
+        corpus_name = check_corpus_name(self.request)
+        if corpus_name:
+            filters.update({
+                'tagged_docs__corpus__name': corpus_name
+            })
+
+        queryset = self.queryset.filter(**filters).distinct()
         serializer = self.get_serializer(queryset, many=True)
         return Response(serializer.data)
+
+class DocumentTagsView(APIView):
+    permission_classes = [IsAuthenticated, CorpusAccessPermission]
+
+    def get(self, request, *args, **kwargs):
+        '''
+        Get the tags for a document
+        '''
+
+        doc = self._get_document(**kwargs)
+
+        if doc:
+            tags = doc.tags.filter(user=request.user)
+        else:
+            tags = []
+
+        serializer = TagSerializer(tags, many=True)
+        return Response(serializer.data)
+
+    def patch(self, request, *args, **kwargs):
+        '''
+        Add or remove tags for a document
+
+        The payload should specify a list of operations, like so:
+
+        ```
+        [
+            {"op": "add", "value": 47},
+            {"op": "remove": "value": 123},
+            {"op": "add", "value": 12},
+        ]
+        ```
+        '''
```

Comment:
Hm, perhaps this could be changed to the frontend sending the updated tags for a document, though they would only be the tags _for that user_.

Reply:
Now that I think more on it, `PATCH` should not actually fully replace a resource. That's what `PUT` is for. Still, only patching a single field should still ideally contain the updated field as a payload, not an 'extra' syntax. I'll see if I can build it like that before merging.

Reply:
Yes, that's what I would imagine here; a payload like `{ "tags": 47, 12 }`

Reply:
This would affect #1199 , which is implementing the view as it is now. But it should not be a problem to update it.

Reply:
Okay, I've changed it so you just send the updated list of tags. The return data now also represents the TaggedDocument object rather than a list of tags.